<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">
  <gazebo>
   <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
     <robotNamespace>/arm_gimbal</robotNamespace>
   </plugin>
  </gazebo>

  <xacro:property name="PI" value="3.1415926535897931" />
  <xacro:property name="PI2" value="${PI/2}" />
  <xacro:property name="PI4" value="${PI2/2}" />

  <xacro:macro name="gimbal" params="name parent *origin">
    <link name="p_base">
        <collision>
          <origin xyz="0 0 0" rpy="0 0 0"/>
          <geometry>
            <box size="0.02 0.02 0.004"/>
          </geometry>
        </collision>

        <inertial>
          <origin xyz="0 0 0" rpy="0 0 0"/>
          <mass value="0.01"/>
          <inertia ixx="0.01"  ixy="0"  ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
        </inertial>

        <visual>
          <geometry>
            <box size="0.02 0.02 0.004"/>
          </geometry>
          <material name="blue">
            <color rgba="0 0 1 0.5"/>
          </material>
        </visual>
    </link>
 <gazebo reference="p_base">
            <material>Gazebo/Grey</material>
        </gazebo>

   <!-- <link name="top_link">
      <collision>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <geometry>
          <box size="0.01 0.01 0.01"/>
        </geometry>
      </collision>

      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="0.01"/>
        <inertia ixx="0.01"  ixy="0"  ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
      </inertial>

      <visual>
        <geometry>
          <mesh filename="package://dji_gimbal/meshes/top.dae"/>
        </geometry>
        <material name="red">
          <color rgba="1 0 0 0.5"/>
        </material>
      </visual>
    </link>-->

    <link name="p_back_link">
      <collision>
        <geometry>
          <box size="0.01 0.01 0.01"/>
        </geometry>
      </collision>

      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="0.01"/>
        <inertia ixx="0.01"  ixy="0"  ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
      </inertial>

      <visual>
        <geometry>
          <box size="0.007 0.025 0.007"/>
        </geometry>
        <origin xyz="0 0 0.0125" rpy="${-PI2} 0 ${PI2}"/>
        <material name="red">
          <color rgba="1 0 0 0.5"/>
        </material>
      </visual>
    </link>

    <link name="p_roll_link">
      <collision>
        <geometry>
          <box size="0.03 0.03 0.03"/>
        </geometry>
      </collision>

      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="0.01"/>
        <inertia ixx="0.01"  ixy="0"  ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
      </inertial>
    </link>
       <gazebo reference="p_roll_link">
            <material>Gazebo/Blue</material>
        </gazebo>

    <link name="p_left_link">
      <collision>
        <geometry>
          <box size="0.01 0.01 0.01"/>
        </geometry>
      </collision>

      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="0.01"/>
        <inertia ixx="0.01"  ixy="0"  ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
      </inertial>

    
    </link>

    <link name="p_right_link">
      <collision>
        <geometry>
          <box size="0.01 0.01 0.01"/>
        </geometry>
      </collision>

      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="0.01"/>
        <inertia ixx="0.01"  ixy="0"  ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
      </inertial>

     
    </link>

    <link name="p_cam_pad">
      <collision>
        <geometry>
          <box size="0.01 0.01 0.01"/>
        </geometry>
      </collision>

      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="0.01"/>
        <inertia ixx="0.01"  ixy="0"  ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
      </inertial>
    </link>
        <gazebo reference="p_cam_pad">
            <material>Gazebo/DarkGrey</material>
        </gazebo>

<link name="platform_link">
 <collision>
        <geometry>
          <box size="0.073 0.073 0.007"/>
        </geometry>
      </collision>

      <inertial>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <mass value="0.1"/>
        <inertia ixx="0.01"  ixy="0"  ixz="0" iyy="0.01" iyz="0" izz="0.01"/>
      </inertial>

      <visual>
        <geometry>
          <box size="0.073 0.073 0.007"/>
        </geometry>
        <origin xyz="0 0 0" rpy="0 0 0"/>
        <material name="red">
          <color rgba="1 0 0 0.5"/>
        </material>
      </visual>
    </link>
        <gazebo reference="platform_link">
            <material>Gazebo/DarkGrey</material>
        </gazebo>

  <joint name="platfrom_joint" type="fixed">  
      <parent link="p_cam_pad"/>
      <child link="platform_link"/>
      <origin xyz="0.0 -0.0365 0" rpy="0 0 0"/>
    </joint>   

    <joint name="iris_joint" type="fixed">
      <xacro:insert_block name="origin" />
      <parent link="${parent}"/>
      <child link="p_base"/>
    </joint>
        
  <!--  <joint name="yaw" type="continuous">
<limit effort="1000.0" velocity="10"/>
      <parent link="base"/>      
      <child link="top_link"/>
      <axis xyz="0 0 1"/>
      <origin xyz="0 0 0.005"/>
    </joint>-->

    <joint name="p_top_to_back" type="fixed">
      <parent link="p_base"/>
      <child link="p_back_link"/>    
      <origin xyz="0.015 0 0.0"/>
    </joint>

    <joint name="p_roll" type="continuous">
<limit effort="1000.0" velocity="10"/>
      <parent link="p_back_link"/>
      <child link="p_roll_link"/>
      <axis xyz="1 0 0"/>
      <origin xyz="-0.010 0.0 0.023"/>
    </joint>

    <joint name="p_roll_to_leftsupport" type="fixed">
     
      <parent link="p_roll_link"/>
      <child link="p_left_link"/>    
      <origin xyz=".001 -0.037 -.001"/>
    </joint>

    <joint name="p_roll_to_rightsupport" type="fixed">
      
      <parent link="p_roll_link"/>
      <child link="p_right_link"/>    
      <origin xyz=".001 0.037 -.001"/>
    </joint>

    <joint name="p_pitch" type="continuous">
<limit effort="10000.0" velocity="100"/>
      <parent link="p_right_link"/>
      <child link="p_cam_pad"/>
      <axis xyz="0 1 0"/>
      <origin xyz="-0.0375 0 0"/> 
    </joint>
  </xacro:macro>

  <!--<transmission name="tran_yaw">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="yaw">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor1">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>-->

  <transmission name="tran_p_pitch">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="p_pitch">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor2">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="tran_p_roll">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="p_roll">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="motor3">
      <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
</robot>
